tasmota/xnrg_22_bl6523.ino

```diff
@@ -0,0 +1,362 @@
+ *    BL6523 TX ->1KOhm->  ESP IO5(D1) (Should be Input Capable)
+ *    BL6523 GND -> ESP GND
+ * 
+ * Connection Eg (ESP32) - Non - Isolated:
+ *    BL6523 RX ->1KOhm-> ESP IO4 (Should be Input Capable)
+ *    BL6523 TX ->1KOhm-> ESP IO5 (Should be Input Capable)
+ *    BL6523 GND -> ESP GND
+ * 
+ * To build add the below to user_config_override.h
+ * #define USE_ENERGY_SENSOR  // Enable Energy sensor framework
+ * #define USE_BL6523         // Add support for Chinese BL6523 based Watt hour meter (+1k code)¸
+ * 
+ * After Installation use the below template sample:
+ * {"NAME":"BL6523 Smart Meter","GPIO":[0,0,0,0,7488,7520,0,0,0,0,0,0,0,0],"FLAG":0,"BASE":18}
+\*********************************************************************************************/
+
+#define XNRG_22 22
+
+#include <TasmotaSerial.h>
+
+#define BL6523_RX_DATASET_SIZE 2
+#define BL6523_TX_DATASET_SIZE 4
+
+#define BL6523_BAUD 4800
+#define BL6523_REG_AMPS 0x05
+#define BL6523_REG_VOLTS 0x07
+#define BL6523_REG_FREQ  0x09
+#define BL6523_REG_WATTS 0x0A
+#define BL6523_REG_POWF  0x08
+#define BL6523_REG_WATTHR 0x0C
+
+#define SINGLE_PHASE 0
+
+/* No idea how to derive human readable units from the byte stream.
+For now  dividing the 24-bit values with below constants seems to yield something sane
+that matches whatever displayed in the screen of my 240v model. Probably it would be possible
+to extract these values from the write register commands (0xCA).*/ 
+//#define BL6523_DIV_AMPS 297899.4f
+//#define BL6523_DIV_VOLTS 13304.0f
+//#define BL6523_DIV_FREQ  3907.0f
+//#define BL6523_DIV_WATTS 707.0f
+//#define BL6523_DIV_WATTHR 674.0f
+
+#define BL6523_IREF 297899
+#define BL6523_UREF 13304
+#define BL6523_FREF 3907
+#define BL6523_PREF 707
+#define BL6523_PWHRREF_D 33 // Substract this from BL6523_PREF to get WattHr Div.
+
+TasmotaSerial *Bl6523RxSerial;
+TasmotaSerial *Bl6523TxSerial;
+
+struct BL6523
+{
+  uint8_t type = 1;
+  uint8_t valid = 0;
+  uint8_t got_data_stone = 0;
+  bool discovery_triggered = false;
+} Bl6523;
+
+bool Bl6523ReadData(void)
+{
+   uint32_t powf_word = 0, powf_buf = 0;
+   float powf = 0.0f;
+
+  if (!Bl6523RxSerial->available())
+  {
+    AddLog(LOG_LEVEL_DEBUG, PSTR("BL6:No Rx Data available" ));
+    return false;
+  }
+
+  while ((Bl6523RxSerial->peek() != 0x35) && Bl6523RxSerial->available())
+  {
+    Bl6523RxSerial->read();
+  }
+
+  if (Bl6523RxSerial->available() < BL6523_RX_DATASET_SIZE)
+  {
+    AddLog(LOG_LEVEL_DEBUG, PSTR("BL6:Rx less than expected" ));
+    return false;
+  }
+
+  uint8_t rx_buffer[BL6523_RX_DATASET_SIZE];
+  Bl6523RxSerial->readBytes(rx_buffer, BL6523_RX_DATASET_SIZE);
+  Bl6523RxSerial->flush(); // Make room for another burst
+
+  AddLogBuffer(LOG_LEVEL_DEBUG_MORE, rx_buffer, BL6523_RX_DATASET_SIZE);
+  
+  while (Bl6523TxSerial->available() < BL6523_TX_DATASET_SIZE)
+  {
+    // sleep till TX buffer is full
+     unsigned long timeout = millis() + 10;
+     while (millis() < timeout) {}
+  }
```

Comment:
Do a ``delay(10);`` here instead of both ``timeout`` lines. This allows for background wifi updates.

Make an abort from the while loop too in case serial data is never received.

Reply:
Indeed a good catch. Thanks